### DISCUSSION TITLE
Trying to understand the source code generated

### DISCUSSION THREAD

**Discussion by ThorvaldAagaard** [NationalSecurityAgency/ghidra#5399] [Q&A]
I had a question about compiling the generated source, and learned that I had to make some changes #236 - that is understandable, but now looking at the generated source I am missing the definition of a lot of fields (or constants), so how was it intended that I should get these definitions

Let me show an example

During startup of the application the following function is called
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/3d2819cf-e6f2-430c-831d-e9d26ac0f9fc)

Reading the documentation for the program, this is some global constants, that are being initialized

The generated c/c++ code then contains this function
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/e8af89a5-3823-4ccf-8b00-fd2774dad685)

And that is very fine, 

But I am missing the defintions of all the variables DAT_*

I can in the code see, where they are assign an initial value
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/1f72b353-9308-4f41-afba-61b9136a3210)

Isn't it possible for Ghidra to generate the missing definition?

I have also noticed, that there are a lot of different CONCATXX- usage, also without any definition, like this
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/ebc617dc-20af-4490-bad7-f4c801364bea)

Where can I find the meaning of that function?

And then I have some starnge assignments
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/f5bd75df-816c-4a10-9778-2e5494b43fc9)

**Comment by ThorvaldAagaard** [comment]
So after adding all the missing definitions (about 1200), I hit the next problem.

1: A lot of functions is returning void, but still assigned in the code
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/cb32cb03-077f-4cea-930a-d9697ac43c9b)

Calling a function returning void in an if-expression seems strange

2: A lot of methods have more parameters, than when actually is passed to the function
An example
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/ab585d14-b0f4-4459-9778-a31d21c0fec6)
Here 2 parameters are passed and looking at the listing it seems OK
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/8711e977-d982-4fb0-9c7f-188703b0a4d2)

But  then looking at the function called
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/52e610af-d6ec-4dd8-996a-8a1b7ef8f1a5)
and from the listing it seems to pop 5 values from the Stack
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/785f5be6-cde8-442e-a622-22efce7087fc)

3: A lot of code references to struct not defined when the source was exported needs to be fixed
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/89f4e49b-ad61-4510-b8cb-d1e6760c70cd)

So here I will have to create the structure, but Ghidra could have added the attribute names for me

4: I haven't figured out how to handle the many different version of CONCATxx

**Reply by astrelsky** [reply]
The first few "issues" you are seeing here are because you have not committed the function parameters. Until you have done so, the decompiler will attempt to determine what parameters the function takes. This depends on stack/register usage and the prototype evaluation setting (I think that is the setting name). Since the context surrounding the function calls is different each time, you will see the behavior you have described.

In the auto analysis options there is a decompiler parameter id analyzer. This analyzer has an option to commit void return types which is off by default. For your use case you might want to enable it.

I **highly** recommend doing a lot of manual analysis prior to what you are attempting to do. Especially when Ghidra isn't able to automatically identify functions such as `printf`.

**Comment by ThorvaldAagaard** [comment]
5: The following code looks like an error in the generator
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/68b28421-eb1f-495b-a991-a17aebba206f)
The function FUN_00405cc0 just prints the text to standard out, and is working fine here

But when you try to format the string in the call Ghidra will generate something like
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/665b7950-abad-4d51-86da-c8922b57ab3b)

My guess is that the function call to create the parameter is lost, and the code probably should be something like
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/af647011-243f-40d4-a7b4-d5ee6fdbf020)

but in gcc it seems sprint does not return the string

**Reply by astrelsky** [reply]
From looking at your screenshots here, FUN_00405cc0 is most likely `printf`

**Reply by ThorvaldAagaard** [reply]
Yes, I concluded the same, but what is interesting is that in the generated source code I can see use of functions like _malloc, so why did Ghidra not detect the use of ```fprintf```

You wrote: _In the auto analysis options there is a decompiler parameter id analyzer. This analyzer has an option to commit void return types which is off by default._ 
but I am not sure where to change this, as when I import a file, this is the options for the analyzer
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/d9b3337c-7472-41bd-8d16-af51a9594421)

But searching for the options I noticed the "Parse C Source" with this configuration
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/cf574840-2297-4013-9ad7-f2ea4c6da8d4)

I have select Visual Studio as I assume the application is built with an old version of visual studio.

Here it seems it could be possible to tweak the generation in multiple ways, but I found this video describing how to use it
https://www.youtube.com/watch?v=u15-r5Erfnw

So in the Data Type Manager I could see that´´´fprintf is available´´´
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/6af8645d-3681-4380-a0bd-926ac2a4c5d2)

So now I just have to find a way to tell Ghidra that FUN_00405cc0 should be replaced with fprintf

**Reply by ThorvaldAagaard** [reply]
Trying to figure out how to handle this I noticed in my Symbol Tree
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/0865f4ae-4235-4811-97a9-d85356910d70)
that there are functions functions marked with FID_conflict and that could be the problem, but I found nothing in the help about this kind of conflict

**Reply by astrelsky** [reply]
> Yes, I concluded the same, but what is interesting is that in the generated source code I can see use of functions like _malloc, so why did Ghidra not detect the use of `fprintf`
> 
> You wrote: _In the auto analysis options there is a decompiler parameter id analyzer. This analyzer has an option to commit void return types which is off by default._ but I am not sure where to change this, as when I import a file, this is the options for the analyzer ![image](https://user-images.githubusercontent.com/10560652/243101551-d9b3337c-7472-41bd-8d16-af51a9594421.png)
> 

Those are the importer options. After importing the file and opening it in the code browser it'll ask if you want to analyze the file. When you say yes all the analysis options will appear.

**Reply by ThorvaldAagaard** [reply]
I have found the option and tested it. At first I like that option, because it looked like it was much better using normal types, and there was a lot of type casting, that no longer was generated. 
An example
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/5a4dbefc-40e4-4d2c-a196-0185ddb15249)
and with the option enabled

![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/145ddef6-bf5e-42b9-9862-97611a5ef200)


But it did not solve the problem with ```fprintf``` and even made it worse as can be seen on this example
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/318382c1-4a4b-42aa-b9d0-d1c28f5292e1)
After enabling the option it became
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/739ef1e4-4dd0-4b8d-8610-2f567564d704)
so the parameters just vanished.

Now inspecting the generated code I think my example wasn't for ```fprintf``` but for ```fwprintf``` but the problem is the same, and looking at the code here it seems obvious there is some conflict between the generated function and fwprintf

![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/776e4294-cefe-4938-8c30-25991baa8ff5)

I followed the implementation of this, and got to
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/9c187207-b453-4dfe-9fe5-885060c01a3b)

And there are 12 of these in the generated code, and it is all the functions that I would have expected Ghidra to identify.

I could probably usie this information to manually edit the generated source code, but still looking for a way getting Ghidra to detect fprintf correctly

**Reply by ThorvaldAagaard** [reply]
I have written some small programs with the code as I expected it to look like, and here Ghidra is identifying fprintf correctly, so I am back to looking at the problem with Multiple Matches and conflicts as the root cause to the generation problems.

Any hints about how to resolve the problem with multiple Matches?

**Comment by topia** [comment]
Did you read `Help` > `Ghidra Functionality` > `Decompiler` > `Decompiler Concepts` and `Program Annotations Affecting the Decompiler`?
I think it help you.

**Comment by ThorvaldAagaard** [comment]
I will try to read the help more carefully, but on first sight it did not help me to generate the definitions for all global variables.

I am also thinking about my problem with no 5, and that could be that Ghidra just decompiled printf, so I should remove the function and replace the call with printf - That would make sense to me.

I did try to find some help about the CONAT, but this is all there is in help:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/b01630ab-b14f-478d-859d-389df5e4da83)

So I still have to figure out how to handle those :-)

**Comment by topia** [comment]
In my understanding, "Decompiler" is the function decompiler currently. No feature about global variables, structures, and so on.

**Comment by topia** [comment]
> I did try to find some help about the CONAT, but this is all there is in help:

You should find the `Internal Decompiler Functions` section.
![image](https://github.com/NationalSecurityAgency/ghidra/assets/13070/bea0ba4b-c0fc-4949-ae89-3a4713fb4f50)

**Comment by ThorvaldAagaard** [comment]
Well the c-decompile generates a .h file with some declarations, and the Code Browser have identified all the locations for global variables, and does a great job generating variables in all functions, so the step to generate the last should not be that big.
And thx for explaining CONCAT, now I just need to implement it :-)
